### PR TITLE
CSCwa43372 - Adding Support for file uploads via rest API

### PR DESCRIPTION
**Pull request by phdawson**
Creating DoChunk function to allow client to accept and send incomplete payloads. This will be used for uploading images to the APIC firmware repository in chunks.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -565,6 +565,57 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
+				return nil, errors.New(fmt.Sprintf("Failed to connect to APIC. Verify that you are connecting to an APIC.\nError message: %+v", err))
+			} else {
+				log.Printf("[ERROR] HTTP Connection failed: %s, retries: %v", err, attempts)
+				continue
+			}
+                }
+
+		bodyBytes, err := ioutil.ReadAll(resp.Body) //this should be fine tho lol
+		bodyStr := string(bodyBytes)
+                fmt.Println("big req resp" , bodyStr)
```
Comment (jjagadis) [inline client/client.go:591]: Please remove this

**Comment by sajagana** [comment]
Please fix the formatting issue with "gofmt".

File: client/client.go [hunk 2]
```diff
@@ -565,6 +565,56 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
 	}
 }
 
+func (c *Client) DoBig(req *http.Request) (*http.Response, error) {
+	// or go home (?)
```
Comment (jjagadis) [inline client/client.go:569]: Do you plan to keep this comment ?

File: client/client.go [hunk 3]
```diff
@@ -565,6 +565,56 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
+	log.Printf("[DEBUG] Begining DoBig method %s", req.URL.String())
+
+	// retain the request body across multiple attempts
+	for attempts := 0; ; attempts++ {
+		log.Printf("[TRACE] HTTP Request Method and URL: %s %s", req.Method, req.URL.String())
+		resp, err := c.httpClient.Do(req)
+		if err != nil {
+			if ok := c.backoff(attempts); !ok {
+				log.Printf("[ERROR] HTTP Connection error occured: %+v", err)
+				log.Printf("[DEBUG] Exit from Do method")
```
Comment (jjagadis) [inline client/client.go:580]: Multiple log statement have "Do method", good idea to change the log to match function name

File: client/client.go [hunk 4]
```diff
@@ -565,6 +565,56 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
+			} else {
+				log.Printf("[ERROR] HTTP Connection failed: %s, retries: %v", err, attempts)
+				continue
+			}
+		}
+
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if (resp.StatusCode < 500 || resp.StatusCode > 504) && resp.StatusCode != 405 {
+			//obj, err := container.ParseJSON(bodyBytes)
```
Comment (xinyuezhao) [inline client/client.go:591]: may want to remove this line if not needed

File: client/client.go [hunk 5]
```diff
@@ -565,6 +565,56 @@ func (c *Client) Do(req *http.Request) (*container.Container, *http.Response, er
...
+				log.Printf("[ERROR] Error occured while json parsing %+v", err)
+				log.Printf("[DEBUG] Exit from Do method")
+				return resp, errors.New(fmt.Sprintf("Failed to parse JSON response from %s. Verify that you are connecting to an APIC.\nHTTP response status: %s\nMessage: %+v", req.URL.String(), resp.Status, err))
+			}
+
+			log.Printf("[DEBUG] Exit from Do method")
+			return resp, nil
+		} else {
+			if ok := c.backoff(attempts); !ok {
+				//obj, err := container.ParseJSON(bodyBytes)
```
Comment (xinyuezhao) [inline client/client.go:602]: may want to remove this line if not needed

**Review by xinyuezhao — CHANGES_REQUESTED** [review]

**Comment by lhercot** [comment]
As we have solved this issue with adding a condition in the existing methods in case retries = 0, I am closing this PR.

**Pull request closed without merging**